core/corehttp/hostname.go

```diff
@@ -151,16 +159,44 @@ func HostnameOption() ServeOption {
 					return
 				}
 
-				// Do we need to fix multicodec in PeerID represented as CIDv1?
-				if isPeerIDNamespace(ns) {
-					keyCid, err := cid.Decode(rootID)
-					if err == nil && keyCid.Type() != cid.Libp2pKey {
-						if newURL, ok := toSubdomainURL(hostname, pathPrefix+r.URL.Path, r); ok {
-							// Redirect to CID fixed inside of toSubdomainURL()
+				// Check if rootID is a valid CID
+				if rootCID, err := cid.Decode(rootID); err == nil {
+					// Do we need to redirect root CID to a canonical DNS representation?
+					dnsCID, err := toDNSPrefix(rootID, rootCID)
+					if err != nil {
```

Comment:
This is purely because of Security Model in modern browsers.

**tl;dr** We can't have path-based and subdowmain-based gateways **on the same hostname** because that would defeat the purpose of Origin isolation provided by the subdomain gw. Namely, `example.com/ipfs/{long-cid}/index.html` could manage permissions/cookies/storage of every CID loaded  under `*.ipfs.example.com` 


Refs.
- https://developer.mozilla.org/en-US/docs/Web/Security/Same-origin_policy
- https://web.dev/same-origin-policy/

---

That being said, we could modify behavior proposed in this PR and give a configuration option to **redirect to path-based gateway on other hostname** instead of returning Error 400 Bad Request. But that is something we should do in separate PR, I believe.

Reply:
Ah ok, seems reasonable. Although I guess if we wanted to we could reserve a subdomain so we have `cidtoolong.ipfs.example.com/{long-cid}/index.html`. You're right that we could do this in a subsequent PR though